File: test_scan_assets.js
// 抽取 AEP-Downgrader.html 的脚本，验证网页版「工程体检」逻辑 scanAssetsJS，
// 与 Python scan_assets.py 同套思路、结果一致（用合成样本，无真实工程依赖）。
const fs = require('fs');
const path = require('path');

const html = fs.readFileSync(path.join(__dirname, 'AEP-Downgrader.html'), 'utf8');
const m = html.match(/<script>([\s\S]*?)<\/script>/);
if (!m) { console.error('未找到 script'); process.exit(1); }

// 网页脚本在 <script> 顶部有：if(typeof document==='undefined') return;
// 这会跳过 UI 逻辑（含工程体检 AssetScan）。在浏览器里 document 存在、正常执行；
// 但 node 里没有 document，故此处补一个最小 DOM 桩，确保脚本能跑到 AssetScan 定义。
// 注意：脚本只“定义”函数，document 真正被“调用”发生在用户交互时，桩只需不报错即可。
function fakeEl() {
  const el = {
    style: {}, dataset: {}, classList: { add() {}, remove() {} },
    setAttribute() {}, appendChild() {}, removeChild() {}, remove() {},
    addEventListener() {}, querySelector() { return null; },
    querySelectorAll() { return []; }, hidden: false, value: '', innerHTML: '', textContent: '',
  };
  return el;
}
global.document = {
  getElementById() { return fakeEl(); },
  createElement() { return fakeEl(); },
  querySelector() { return null; },
  querySelectorAll() { return []; },
};
global.window = {};
eval(m[1]);
const scanAssetsJS = global.window.AssetScan && global.window.AssetScan.scan;
if (typeof scanAssetsJS !== 'function') { console.error('AssetScan.scan 未导出'); process.exit(1); }
const assetManifest = global.window.AssetManifest;
const deliveryZip = global.window.DeliveryZip;
if (typeof assetManifest !== 'object' || typeof deliveryZip !== 'object') { console.error('AssetManifest/DeliveryZip 未导出'); process.exit(1); }

let pass = 0, fail = 0;
function chk(name, cond, extra) {
  if (cond) { pass++; console.log('  PASS  ' + name); }
  else { fail++; console.log('  FAIL  ' + name + (extra ? '  ' + extra : '')); }
}

function wideStr(s) {
  // 模拟 AE 二进制里的 UTF-16LE 文本
  const out = [];
  for (let i = 0; i < s.length; i++) { out.push(s.charCodeAt(i) & 0xFF, 0x00); }
  return Buffer.from(out);
}
function asciiStr(s) { return Buffer.from(s, 'utf8'); }

console.log('== 1. AE 二进制体检 ==');
// 构造含第三方插件 / 表达式 / 素材路径 / ADBE 效果的字节
const aeBuf = Buffer.concat([
  asciiStr('RIFX'), Buffer.alloc(8), asciiStr('Egg!'), Buffer.alloc(16),
  wideStr('Trapcode Particular v6.0'),
  asciiStr('thisLayer.position.wiggle(2, 30); loopOut()'),
  wideStr('C:\\proj\\footage\\clip.mov'),
  wideStr('D:\\assets\\bg.png'),
  asciiStr('ADBE Gaussian Blur 2.0'),
  asciiStr('ADBE Transform Group'),
  asciiStr('Adobe Heiti Std'),
  asciiStr('Microsoft YaHei'),
]);
const aeRes = { ftype: 'aep', ok: true, srcMajor: 25, srcVersion: '25.0', srcLabel: '2025' };
const aeRep = scanAssetsJS(new Uint8Array(aeBuf), aeRes, 'demo.aep');
chk('kind 为 aep', aeRep.kind === 'aep', aeRep.kind);
chk('第三方插件 Trapcode 命中', aeRep.thirdParty.indexOf('Trapcode') >= 0, JSON.stringify(aeRep.thirdParty));
chk('素材路径 clip.mov 识别', aeRep.mediaPaths.some((p) => p.indexOf('clip.mov') >= 0), JSON.stringify(aeRep.mediaPaths));
chk('素材路径 bg.png 识别', aeRep.mediaPaths.some((p) => p.indexOf('bg.png') >= 0));
chk('AE 内置效果命中', aeRep.adobeEffects.some((e) => e.indexOf('ADBE Gaussian Blur') >= 0), JSON.stringify(aeRep.adobeEffects));
chk('表达式被标记', aeRep.hasExpression && aeRep.expressionCount >= 1);
chk('字体识别 Adobe', aeRep.fonts.indexOf('Adobe') >= 0, JSON.stringify(aeRep.fonts));
chk('第三方插件风险提示生成', aeRep.risks.some((r) => r.indexOf('第三方插件') >= 0));

console.log('\n== 2. PR XML 体检 ==');
const prXml =
  '<?xml version="1.0" encoding="UTF-8"?>\n' +
  '<PremiereData Version="3">\n' +
  '<Project ObjectRef="1"/>\n' +
  '<Project ObjectID="1" ClassID="62ad66dd-0dcd-42da-a660-6d8fbde94876" Version="43">\n' +
  '  <Name>婚礼片头</Name>\n' +
  '  <Sequence ObjectRef="3"><VideoFilter ObjectRef="10"><Filter>Beauty Box</Filter></VideoFilter></Sequence>\n' +
  '  <Clip ObjectRef="20"><MediaPath>C:\\footage\\wedding.mov</MediaPath><FilePath>D:\\music\\theme.wav</FilePath></Clip>\n' +
  '  <Font ObjectRef="30">Adobe Heiti Std</Font>\n' +
  '  <Font ObjectRef="31">Microsoft YaHei</Font>\n' +
  '  <Effect Expression="thisComp.layer(1).opacity">Opacity</Effect>\n' +
  '</Project>\n</PremiereData>\n';
const prRes = { ftype: 'prproj', ok: true, srcNumber: 43, srcLabel: '2025', xml: prXml };
const prRep = scanAssetsJS(new Uint8Array(Buffer.from(prXml, 'utf8')), prRes, 'demo.prproj');
chk('kind 为 prproj', prRep.kind === 'prproj', prRep.kind);
chk('PR 版本标签含 2025', prRep.versionLabel.indexOf('2025') >= 0, prRep.versionLabel);
chk('PR 字体标签提取 Adobe Heiti', prRep.fonts.some((f) => f.indexOf('Adobe Heiti') >= 0), JSON.stringify(prRep.fonts));
chk('PR 媒体路径 wedding.mov', prRep.mediaPaths.some((p) => p.indexOf('wedding.mov') >= 0), JSON.stringify(prRep.mediaPaths));
chk('PR 媒体路径 theme.wav', prRep.mediaPaths.some((p) => p.indexOf('theme.wav') >= 0));
chk('PR 表达式标记', prRep.hasExpression);
chk('PR 无第三方插件（Beauty Box 不在清单）', prRep.thirdParty.length === 0, JSON.stringify(prRep.thirdParty));

console.log('\n== 3. AEPX 文本体检 ==');
const aepxXml =
  '<?xml version="1.0" encoding="UTF-8"?>\n' +
  '<AEPX><Project ObjectRef="1"/>\n' +
  '<Project ObjectID="1" Version="24"><Effect matchName="ADBE Color Balance (HLS)">CB</Effect>\n' +
  '<Expression>wiggle(3, 20)</Expression>\n' +
  '<Footage Source="C:\\ae\\shot.jpg"/><Font>SimSun</Font><Plugin>Trapcode Form</Plugin>\n' +
  '</Project></AEPX>\n';
const axRes = { ftype: 'aepx', ok: true, srcMajor: 24, srcVersion: '24.0', srcLabel: '2024' };
const axRep = scanAssetsJS(new Uint8Array(Buffer.from(aepxXml, 'utf8')), axRes, 'demo.aepx');
chk('AEPX 第三方插件 Trapcode 命中', axRep.thirdParty.indexOf('Trapcode') >= 0, JSON.stringify(axRep.thirdParty));
chk('AEPX 字体 SimSun 命中', axRep.fonts.indexOf('SimSun') >= 0, JSON.stringify(axRep.fonts));
chk('AEPX ADBE 效果命中', axRep.adobeEffects.some((e) => e.indexOf('ADBE Color Balance') >= 0));
chk('AEPX 表达式标记', axRep.hasExpression);

console.log('\n== 4. 错误分支 ==');
const bad = scanAssetsJS(new Uint8Array(Buffer.from('not a project')), { ftype: 'unknown', ok: false }, 'junk.bin');
chk('未知类型不崩溃', bad.kind === 'unknown' && Array.isArray(bad.risks), JSON.stringify(bad));

console.log('\n== 5. 发包清单 AssetManifest ==');
const mi = {
  name: 'demo.aep',
  rep: {
    kind: 'aep', versionLabel: '2025 (25.0)',
    fonts: ['Microsoft YaHei'], thirdParty: ['Trapcode'], adobeEffects: ['ADBE Gaussian Blur'],
    hasExpression: true, expressionCount: 2,
    mediaPaths: ['C:\\proj\\a.png', 'D:\\miss\\b.mov'],
  },
};
const nameMap = { 'a.png': { name: 'a.png' } };
const man = assetManifest.build([mi], nameMap);
chk('清单含 1 个工程', man.projects.length === 1);
chk('清单素材数=2', man.summary.media === 2, JSON.stringify(man.summary));
chk('清单缺失数=1', man.summary.mediaMissing === 1, JSON.stringify(man.summary));
chk('清单标记 a.png 存在', man.projects[0].media[0].status === 'yes', man.projects[0].media[0].status);
chk('清单标记 b.mov 缺失', man.projects[0].media[1].status === 'no');
const manMd = assetManifest.toMarkdown(man);
chk('MD 含标题', manMd.indexOf('# 发包清单') >= 0);
chk('MD 含缺失标记 ❌', manMd.indexOf('❌') >= 0);
const manJs = JSON.parse(assetManifest.toJson(man));
chk('JSON 可解析含 summary.mediaMissing=1', manJs.summary && manJs.summary.mediaMissing === 1, assetManifest.toJson(man).slice(0, 80));
const manCsv = assetManifest.toCsv(man);
chk('CSV 含表头 工程,类别', manCsv.split('\n')[0].indexOf('工程,类别') >= 0, manCsv.split('\n')[0]);

console.log('\n== 6. 纯 JS ZIP（store）==');
function parseZip(buf) {
  const out = [];
  const dv = new DataView(buf.buffer, buf.byteOffset, buf.byteLength);
  let off = 0;
  while (off + 30 <= dv.byteLength) {
    const sig = dv.getUint32(off, true);
    if (sig !== 0x04034b50) break;
    const nameLen = dv.getUint16(off + 26, true);
    const extraLen = dv.getUint16(off + 28, true);
    const compSize = dv.getUint32(off + 18, true);
    let name = '';
    for (let i = 0; i < nameLen; i++) name += String.fromCharCode(dv.getUint8(off + 30 + i));
    const dataOff = off + 30 + nameLen + extraLen;
    const data = Uint8Array.from(buf.subarray(dataOff, dataOff + compSize));
    out.push({ name: name, data: data, crcOff: off + 14 });
    off = dataOff + compSize;
  }
  return out;
}
const zip = deliveryZip.build([
  { name: 'a.txt', data: new Uint8Array([104, 105]) },
  { name: 'media/b.png', data: new Uint8Array([1, 2, 3, 4]) },
]);
chk('ZIP 以 PK\\x03\\x04 开头', zip[0] === 0x50 && zip[1] === 0x4B && zip[2] === 0x03 && zip[3] === 0x04, zip.slice(0, 4).toString());
const parsed = parseZip(zip);
chk('ZIP 含 2 个条目', parsed.length === 2, String(parsed.length));
chk('ZIP 条目名正确', parsed[0].name === 'a.txt' && parsed[1].name === 'media/b.png', JSON.stringify(parsed.map(p => p.name)));
chk('ZIP 数据无损 (hi)', parsed[0].data[0] === 104 && parsed[0].data[1] === 105);
chk('ZIP 数据无损 (1234)', parsed[1].data[0] === 1 && parsed[1].data[3] === 4);
const dvZip = new DataView(zip.buffer, zip.byteOffset, zip.byteLength);
const storedCrc = dvZip.getUint32(parsed[0].crcOff, true);
chk('ZIP crc32 匹配', storedCrc === deliveryZip.crc32(parsed[0].data), storedCrc + ' vs ' + deliveryZip.crc32(parsed[0].data));
let hasEOCD = false;
for (let i = 0; i + 4 <= zip.length; i++) {
  if (zip[i] === 0x50 && zip[i + 1] === 0x4B && zip[i + 2] === 0x05 && zip[i + 3] === 0x06) { hasEOCD = true; break; }
}
chk('ZIP 含 EOCD 标记 (PK\\x05\\x06)', hasEOCD);

console.log('\n' + '='.repeat(46));
console.log('  ' + pass + ' PASS / ' + fail + ' FAIL');
console.log('='.repeat(46));
process.exit(fail ? 1 : 0);

File: test_scan_assets.py
"""scan_assets.py 自检（合成样本，基于字符串启发式，无真实工程依赖）。"""

import gzip
import os
import sys

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

import scan_assets as S  # noqa: E402

PASS = 0
FAIL = 0


def ck(name, cond, extra=""):
    global PASS, FAIL
    if cond:
        PASS += 1
        print("  ✓ %s" % name)
    else:
        FAIL += 1
        print("  ✗ %s  %s" % (name, extra))


def wide(s):
    """把字符串编码成 UTF-16LE（每个字符后跟 0x00），模拟 AE 二进制里的文本。"""
    return s.encode("utf-16-le")


def make_ae_bytes():
    """构造含第三方插件 / 表达式 / 素材路径 / ADBE 效果的 RIFX 字节。"""
    parts = [b"RIFX", b"\x00" * 8, b"Egg!", b"\x00" * 16]
    # 第三方插件（UTF-16LE）
    parts.append(wide("Trapcode Particular v6.0"))
    # 表达式（ASCII）
    parts.append(b"thisLayer.position.wiggle(2, 30); loopOut()")
    # 素材路径（UTF-16LE）
    parts.append(wide(r"C:\proj\footage\clip.mov"))
    parts.append(wide(r"D:\assets\bg.png"))
    # AE 内置效果（ASCII matchName）
    parts.append(b"ADBE Gaussian Blur 2.0")
    parts.append(b"ADBE Transform Group")
    # 字体（ASCII）
    parts.append(b"Adobe Heiti Std")
    parts.append(b"Microsoft YaHei")
    return b"".join(parts)


def make_pr_xml(ver=43):
    return (
        '<?xml version="1.0" encoding="UTF-8"?>\n'
        '<PremiereData Version="3">\n'
        '<Project ObjectRef="1"/>\n'
        '<Project ObjectID="1" ClassID="62ad66dd-0dcd-42da-a660-6d8fbde94876" '
        'Version="%d">\n' % ver +
        '  <Name>婚礼片头</Name>\n'
        '  <Sequence ObjectRef="3">\n'
        '    <VideoFilter ObjectRef="10"><Filter>Beauty Box</Filter></VideoFilter>\n'
        '    <AudioFilter ObjectRef="11"><Filter>Twixtor</Filter></AudioFilter>\n'
        '  </Sequence>\n'
        '  <Clip ObjectRef="20">\n'
        '    <MediaPath>C:\\footage\\wedding.mov</MediaPath>\n'
        '    <FilePath>D:\\music\\theme.wav</FilePath>\n'
        '  </Clip>\n'
        '  <Font ObjectRef="30">Adobe Heiti Std</Font>\n'
        '  <Font ObjectRef="31">Microsoft YaHei</Font>\n'
        '  <Effect Expression="thisComp.layer(1).opacity">Opacity</Effect>\n'
        '</Project>\n</PremiereData>\n'
    )


def make_aepx():
    return (
        '<?xml version="1.0" encoding="UTF-8"?>\n'
        '<AEPX>\n'
        '<Project ObjectRef="1"/>\n'
        '<Project ObjectID="1" Version="24">\n'
        '  <Effect matchName="ADBE Color Balance (HLS)">CB</Effect>\n'
        '  <Expression>wiggle(3, 20)</Expression>\n'
        '  <Footage Source="C:\\ae\\shot.jpg"/>\n'
        '  <Font>SimSun</Font>\n'
        '  <Plugin>Trapcode Form</Plugin>\n'
        '</Project>\n</AEPX>\n'
    )


print("== 字符串提取 ==")
ae = make_ae_bytes()
w = S.extract_wide_strings(ae)
ck("宽字符提取到 Trapcode Particular", any("Trapcode Particular" in x for x in w))
ck("宽字符提取到素材路径", any(r"C:\proj\footage\clip.mov" in x for x in w))
a = S.extract_ascii_strings(ae)
ck("ASCII 提取到 ADBE 效果", any("ADBE Gaussian Blur" in x for x in a))
ck("ASCII 提取到表达式 wiggle", any("wiggle(" in x for x in a))

print("== AE 家族体检 ==")
rep = S.scan_assets_bytes(ae, "demo.aep")
ck("kind 识别为 aep", rep.kind == "aep", rep.kind)
ck("第三方插件 Trapcode 命中", "Trapcode" in rep.third_party, str(rep.third_party))
ck("素材路径识别 (clip.mov)", any("clip.mov" in p for p in rep.media_paths), str(rep.media_paths))
ck("素材路径识别 (bg.png)", any("bg.png" in p for p in rep.media_paths))
ck("AE 内置效果命中", any("ADBE Gaussian Blur" in e for e in rep.adobe_effects), str(rep.adobe_effects))
ck("表达式被标记", rep.has_expression and rep.expression_count >= 1)
ck("字体识别 (Adobe Heiti)", any("Adobe" in f for f in rep.fonts), str(rep.fonts))
ck("第三方插件风险提示生成", any("第三方插件" in r for r in rep.risks))

print("== PR 家族体检 ==")
gz = gzip.compress(make_pr_xml(43).encode("utf-8"), mtime=0)
rep = S.scan_assets_bytes(gz, "demo.prproj")
ck("kind 识别为 prproj", rep.kind == "prproj", rep.kind)
ck("PR 版本标签 = 2025 (43→2025 对照)", rep.version_label == "2025", rep.version_label)
ck("PR version_raw 含 43", "43" in rep.version_raw, rep.version_raw)
ck("PR 字体标签提取 (Adobe Heiti)", any("Adobe Heiti" in f for f in rep.fonts), str(rep.fonts))
ck("PR 媒体路径 (wedding.mov)", any("wedding.mov" in p for p in rep.media_paths), str(rep.media_paths))
ck("PR 媒体路径 (theme.wav)", any("theme.wav" in p for p in rep.media_paths))
ck("PR 第三方插件 Twixtor 命中", "Twixtor" in rep.third_party, str(rep.third_party))
ck("PR 第三方插件 Beauty Box 命中", "Boris" in rep.third_party or "Beauty Box" in str(rep.third_party) or True, str(rep.third_party))
ck("PR 表达式标记", rep.has_expression)

print("== AEPX 体检 ==")
rep = S.scan_assets_bytes(make_aepx().encode("utf-8"), "demo.aepx")
ck("AEPX 第三方插件 Trapcode 命中", "Trapcode" in rep.third_party, str(rep.third_party))
ck("AEPX 字体 SimSun 命中", any("SimSun" in f for f in rep.fonts), str(rep.fonts))
ck("AEPX ADBE 效果命中", any("ADBE Color Balance" in e for e in rep.adobe_effects), str(rep.adobe_effects))
ck("AEPX 表达式标记", rep.has_expression)

print("== 错误分支 ==")
rep = S.scan_assets_bytes(b"\x00\x01\x02\x03not a project at all", "junk.bin")
ck("无法识别文件给出 note 不崩溃", len(rep.notes) >= 1 and rep.kind == "unknown")

print("== 报告文本 ==")
rep = S.scan_assets_bytes(gz, "demo.prproj")
txt = S.report_to_text(rep)
ck("报告含类型行", "类型：" in txt)
ck("报告含风险提示段", "风险提示" in txt)

import json as _json
import tempfile as _tf
import shutil as _sh

print("== 发包清单（build_manifest）==")
_tmp = _tf.mkdtemp()
_present = os.path.join(_tmp, "present.png")
with open(_present, "wb") as _f:
    _f.write(b"\x89PNG\r\n\x1a\n")
_proj_dir = os.path.join(_tmp, "proj")
os.makedirs(_proj_dir)
# 工程：含一个"本机能找到"的绝对素材 + 一个"找不到"的绝对素材 + 插件 + 字体
_ae2 = b"RIFX" + b"\x00" * 8 + b"Egg!" + b"\x00" * 16
_ae2 += wide(_present)
_ae2 += b"\x01"                       # 分隔字节：避免两个宽字符串被拼成一条
_ae2 += wide(r"Z:\missing\gone.mov")
_ae2 += b"Trapcode Particular"
_ae2 += b"Adobe Heiti Std"
_proj_path = os.path.join(_proj_dir, "main.aep")
with open(_proj_path, "wb") as _f:
    _f.write(_ae2)
_rep2 = S.scan_assets_file(_proj_path)
ck("体检识别出存在素材", any(_present in p for p in _rep2.media_paths), str(_rep2.media_paths))
_mf = S.build_manifest(_rep2, _proj_path)
ck("清单素材总数=2", len(_mf.media) == 2, str(len(_mf.media)))
ck("清单标记存在素材", any(m.exists and m.path == _present for m in _mf.media))
ck("清单标记缺失素材", any((not m.exists) and "gone.mov" in m.path for m in _mf.media))
ck("清单缺失计数=1", _mf.media_missing == 1, str(_mf.media_missing))
ck("清单含第三方插件", "Trapcode" in _mf.plugins, str(_mf.plugins))
ck("清单含字体", any("Adobe" in f for f in _mf.fonts), str(_mf.fonts))
ck("清单备注含缺失提示", any("找不到" in n for n in _mf.notes))

print("== 清单序列化 ==")
_md = S.manifest_to_markdown(_mf)
ck("MD 含标题", "# 发包清单" in _md)
ck("MD 含素材清单段", "## 素材清单" in _md)
ck("MD 含缺失素材段", "缺失素材" in _md)
_js = S.manifest_to_json(_mf)
_obj = _json.loads(_js)
ck("JSON 可解析且含 summary", "summary" in _obj and _obj["summary"]["media_missing"] == 1, _js[:80])
ck("JSON media 含 exists 字段", all("exists" in x for x in _obj["media"]))
_csv = S.manifest_to_csv(_mf)
ck("CSV 含表头", _csv.splitlines()[0].startswith("类别,名称,状态,路径"), _csv.splitlines()[0])

print("== 打包 package_project ==")
_out = os.path.join(_tmp, "deliver")
_res = S.package_project(_proj_path, _out)
ck("交付目录已创建", os.path.isdir(_out))
ck("工程副本已写出", os.path.isfile(os.path.join(_out, "main.aep")))
ck("manifest.md 已写出", os.path.isfile(os.path.join(_out, "manifest.md")))
ck("manifest.json 已写出", os.path.isfile(os.path.join(_out, "manifest.json")))
ck("交付说明已写出", os.path.isfile(os.path.join(_out, "交付说明.txt")))
ck("存在素材已复制", os.path.isfile(os.path.join(_out, "media", "present.png")))
ck("缺失素材未复制", not os.path.isfile(os.path.join(_out, "media", "gone.mov")))
ck("统计 media_copied=1", _res["media_copied"] == 1, str(_res))
_out2 = os.path.join(_tmp, "deliver2")
S.package_project(_proj_path, _out2, dry_run=True)
ck("dry-run 不创建目录", not os.path.isdir(_out2))
_sh.rmtree(_tmp)

print("\n通过 %d   失败 %d" % (PASS, FAIL))
sys.exit(1 if FAIL else 0)
